File: 1697. Checking Existence of Edge Length Limited Paths.cpp
class Solution {
public:
    class DSU {
    public:
        vector<int> parent;
        vector<int> rank;

        DSU(int n) : parent(n), rank(n, 0) {
            for (int i = 0; i < n; i++) parent[i] = i;
        }

        int find(int n) {
            if (parent[n] != n) parent[n] = find(parent[n]);
            return parent[n];
        }

        void union_sets(int x, int y) {
            int p1 = find(x);
            int p2 = find(y);
            if (p1 == p2) return;

            if (rank[p1] < rank[p2]) {
                parent[p1] = p2;
            } else if (rank[p1] > rank[p2]) {
                parent[p2] = p1;
            } else {
                parent[p1] = p2;
                rank[p2]++;
            }
        }
    };

    vector<bool> distanceLimitedPathsExist(int n, vector<vector<int>>& edgeList, vector<vector<int>>& qs) {
        sort(edgeList.begin(), edgeList.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[2] < b[2];
        });

        vector<vector<int>> queries(qs.size(), vector<int>(4));
        for (int i = 0; i < (int)qs.size(); i++) {
            queries[i][0] = qs[i][0];
            queries[i][1] = qs[i][1];
            queries[i][2] = qs[i][2];
            queries[i][3] = i;
        }

        sort(queries.begin(), queries.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[2] < b[2];
        });

        DSU dsu(n);
        vector<bool> ans(queries.size());
        int index = 0;
        for (int i = 0; i < (int)queries.size(); i++) {
            while (index < (int)edgeList.size() && edgeList[index][2] < queries[i][2]) {
                dsu.union_sets(edgeList[index][0], edgeList[index][1]);
                index++;
            }

            ans[queries[i][3]] = (dsu.find(queries[i][0]) == dsu.find(queries[i][1]));
        }

        return ans;
    }
};

File: 1697. Checking Existence of Edge Length Limited Paths.java
class Solution {
    public class DSU{
        int[] parent;
        int[] rank;
        public DSU(int n){
            parent = new int[n];
            rank = new int[n];
            for(int i = 0; i < n; i++) parent[i] = i;
        }

        public int find(int n){
            if(parent[n] != n) return parent[n] = find(parent[n]);
            return n;
        }

        public void union(int x, int y){
            int p1 = find(x);
            int p2 = find(y);
            if(p1 == p2) return;

            if(rank[p1] < rank[p2]){
                parent[p1] = p2;
            }
            else if(rank[p1] > rank[p2]){
                parent[p2] = p1;
            }
            else{
                parent[p1] = p2;
                rank[p2]++;
            }
        }
    }
    public boolean[] distanceLimitedPathsExist(int n, int[][] edgeList, int[][] qs) {
        Arrays.sort(edgeList, (a, b) -> Integer.compare(a[2], b[2]));
        int[][] queries = new int[qs.length][4];
        for(int i = 0; i < qs.length; i++){
            queries[i][0] = qs[i][0];
            queries[i][1] = qs[i][1];
            queries[i][2] = qs[i][2];
            queries[i][3] = i;
        }
        Arrays.sort(queries, (a, b) -> Integer.compare(a[2], b[2]));

        DSU dsu = new DSU(n);
        boolean[] ans = new boolean[queries.length];
        int index = 0;
        for(int i = 0; i < queries.length; i++){
            while(index < edgeList.length && edgeList[index][2] < queries[i][2]){
                dsu.union(edgeList[index][0], edgeList[index][1]);
                index++;
            }

            if(dsu.find(queries[i][0]) == dsu.find(queries[i][1])) ans[queries[i][3]] = true;
            else ans[queries[i][3]] = false;
        }

        return ans;
    }
}